routing.go

```diff
@@ -132,75 +136,66 @@ func (dht *IpfsDHT) GetValue(ctx context.Context, key string, opts ...routing.Op
 	return best, nil
 }
 
-// SearchValue searches for the value corresponding to given Key and streams the results.
-func (dht *IpfsDHT) SearchValue(ctx context.Context, key string, opts ...routing.Option) (<-chan []byte, error) {
+// SearchValueExtended searches for the value corresponding to given Key and streams the results.
+func (dht *IpfsDHT) SearchValueExtended(ctx context.Context, key string, opts ...routing.Option) (<-chan []byte, <-chan []peer.ID, error) {
 	if !dht.enableValues {
-		return nil, routing.ErrNotSupported
+		return nil, nil, routing.ErrNotSupported
 	}
 
 	var cfg routing.Options
 	if err := cfg.Apply(opts...); err != nil {
-		return nil, err
+		return nil, nil, err
 	}
 
-	responsesNeeded := 0
-	if !cfg.Offline {
-		responsesNeeded = getQuorum(&cfg, defaultQuorum)
-	}
-
-	stopCh := make(chan struct{})
-	valCh, lookupRes := dht.getValues(ctx, key, stopCh)
-
-	out := make(chan []byte)
-	go func() {
-		defer close(out)
-		best, peersWithBest, aborted := dht.searchValueQuorum(ctx, key, valCh, stopCh, out, responsesNeeded)
-		if best == nil || aborted {
-			return
+	processors := dhtrouting.GetProcessors(&cfg)
+	if processors == nil {
+		validation := &dhtrouting.ValidationFilter{
+			Key:       key,
+			Validator: dht.Validator,
 		}
 
-		updatePeers := make([]peer.ID, 0, dht.bucketSize)
-		select {
-		case l := <-lookupRes:
-			if l == nil {
-				return
-			}
-
-			for _, p := range l.peers {
-				if _, ok := peersWithBest[p]; !ok {
-					updatePeers = append(updatePeers, p)
-				}
-			}
-		case <-ctx.Done():
-			return
+		quorum := &dhtrouting.CountStopper{
+			MaxCount: dhtrouting.GetQuorum(&cfg),
 		}
 
-		dht.updatePeerValues(dht.Context(), key, best, updatePeers)
-	}()
-
-	return out, nil
-}
+		bestValue := &dhtrouting.BestValueFilterRecorder{
+			Key:           key,
+			Validator:     dht.Validator,
+			PeersWithBest: make(map[peer.ID]struct{}),
+		}
 
-func (dht *IpfsDHT) searchValueQuorum(ctx context.Context, key string, valCh <-chan RecvdVal, stopCh chan struct{},
-	out chan<- []byte, nvals int) ([]byte, map[peer.ID]struct{}, bool) {
-	numResponses := 0
-	return dht.processValues(ctx, key, valCh,
-		func(ctx context.Context, v RecvdVal, better bool) bool {
-			numResponses++
-			if better {
-				select {
-				case out <- v.Val:
-				case <-ctx.Done():
-					return false
-				}
+		processors = []dhtrouting.Processor{validation, quorum, bestValue}
```

Comment:
The current code actually does quorum -> bestValue. The faithful representation of this here would be quorum -> validation -> bestValue.